### PR TITLE
Add support for Gradle version catalog dependencies

### LINKED ISSUE #4521: Incorrect Maven packages reported when scanning build.gradle for an Android app
While scanning F-droid sources at:
https://staging.scancode.io/project/fdroid-source-99ed4069/download_input/org.fdroid.fdroid_1023051_src.tar.gz/
these are "implementation" deps from a gradle build:
```
dependencies {
    implementation project(":libs:download")
    implementation project(":libs:index")
    implementation project(":libs:database")
    implementation libs.androidx.appcompat
    implementation libs.androidx.preference.ktx
    implementation libs.androidx.gridlayout
    implementation libs.androidx.recyclerview
    implementation libs.androidx.vectordrawable
    implementation libs.androidx.constraintlayout
    implementation libs.androidx.lifecycle.livedata.ktx
    implementation libs.androidx.work.runtime
    implementation libs.guava // somehow needed for work-runtime to function

    implementation libs.material

    //noinspection UseTomlInstead
    implementation('com.journeyapps:zxing-android-embedded:4.3.0') { transitive = false }
    implementation libs.zxing.core
    implementation libs.guardianproject.netcipher
    implementation libs.commons.io
    implementation libs.commons.net
    implementation libs.acra.mail
    implementation libs.acra.dialog
    implementation libs.adapterdelegates4
    implementation libs.slf4j.api
    implementation libs.logback.android

    implementation libs.rxjava
    implementation libs.rxandroid

    implementation libs.glide
    implementation(libs.glide.compose)
    annotationProcessor libs.glide.compiler

    implementation libs.okhttp
    implementation libs.bcprov.jdk15to18
    fullImplementation libs.guardianproject.panic
    fullImplementation libs.bcpkix.jdk15to18
    fullImplementation libs.jmdns
    fullImplementation libs.nanohttpd
```

these are reported as:
```
 pkg:maven/
accompanist
pkg:maven/acra
pkg:maven/acra
pkg:maven/adapterdelegates4
pkg:maven/androidx
...
```

which is incomplete or incorrect.

### PR DESCRIPTION
**Pull request by Kaushik-Kumar-CEG**
fixes #4521

### The Issue
The current Gradle parser reports incomplete Maven PURLs (e.g., `pkg:maven/androidx`) when scanning `build.gradle` files that use modern Gradle version catalog references (like `libs.androidx.appcompat`).

### The Approach
This PR adds support for parsing `gradle/libs.versions.toml` to dynamically resolve version catalog aliases into full, accurate Maven coordinates.

* **Grammar Update:** Added the `DEPENDENCY-VERSION-CATALOG` rule to match `libs.x.y.z` patterns in the existing `pygmars` parser
* **Catalog Parsing:** Added `parse_version_catalog()` to locate, read, and parse TOML catalog files
* **Resolution:** Updated `get_dependencies()` to intercept and resolve these catalog references during extraction to generate complete PURLs

**Before Fix:**
image

**After Fix:**
image

### Verification
- [x] Ran full test suite for the gradle parser (all tests pass).
- [x] Verified existing Groovy and Kotlin tests are unbroken.
- [x] Tested edge cases: missing catalogs, unresolved references, and mixed dependencies.
- [x] Added new test fixtures (`fdroid-version-catalog`) including a sample `libs.versions.toml` and expected JSON output.

### CODE REVIEW

**Comment by Kaushik-Kumar-CEG** [comment]
@pombredanne @AyanSinhaMahapatra 
reopened the PR! closed it briefly to double check something locally 
Ready for review :)

File: src/packagedcode/build_gradle.py [hunk 1]
```diff
@@ -322,11 +346,81 @@ def get_dependencies_from_parse_tree(parse_tree):
...
+                version = lib_spec.get('version', '')
+                if isinstance(version, dict) and 'ref' in version:
+                    version = versions.get(version['ref'], '')
+                alias_map[normalized_alias] = {
+                    'namespace': group,
+                    'name': name,
+                    'version': version
+                }
+        return alias_map
+    except Exception as e:
```
Comment (AyanSinhaMahapatra) [inline src/packagedcode/build_gradle.py:395]: Exception handling is present upstream in the package plugin as generic catchalls for failure to parse so this is not required.

File: src/packagedcode/build_gradle.py [hunk 2]
```diff
@@ -322,11 +346,81 @@ def get_dependencies_from_parse_tree(parse_tree):
...
+    catalog_locations = [
+        os.path.join(gradle_dir, 'gradle', 'libs.versions.toml'),
+        os.path.join(os.path.dirname(gradle_dir), 'gradle', 'libs.versions.toml'),
+    ]
+    catalog_path = None
+    for loc in catalog_locations:
+        if os.path.exists(loc):
+            catalog_path = loc
+            break
+    if not catalog_path:
```
Comment (AyanSinhaMahapatra) [inline src/packagedcode/build_gradle.py:365]: Could you have spaces between blocks of code, improves readability.

File: src/packagedcode/build_gradle.py [hunk 3]
```diff
@@ -322,11 +346,81 @@ def get_dependencies_from_parse_tree(parse_tree):
...
+            if alias in catalog:
+                catalog_entry = catalog[alias]
+                resolved_dependencies.append({
+                    'namespace': catalog_entry['namespace'],
+                    'name': catalog_entry['name'],
+                    'version': catalog_entry['version'],
+                    'scope': dep.get('scope', '')
+                })
+            # Skip if not found in catalog - prevents incomplete PURLs
+            continue
```
Comment (AyanSinhaMahapatra) [inline src/packagedcode/build_gradle.py:420]: Not sure if I want to ignore these altogether, I still want to have these as before, shows something is wrong instead of ignoring silently.

File: tests/packagedcode/data/build_gradle/groovy/fdroid-version-catalog/build.gradle-expected.json [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+      },
+      {
+        "purl": "pkg:maven/androidx.appcompat/appcompat@1.6.1",
+        "extracted_requirement": "1.6.1",
+        "scope": "implementation",
+        "is_runtime": true,
+        "is_optional": false,
+        "is_pinned": true,
+        "is_direct": true,
+        "resolved_package": {},
```
Comment (AyanSinhaMahapatra) [inline tests/packagedcode/data/build_gradle/groovy/fdroid-version-catalog/build.gradle-expected.json:59]: We want to create a `resolved_package` for pinned requirements.

File: tests/packagedcode/data/build_gradle/groovy/fdroid-version-catalog/gradle/libs.versions.toml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+[versions]
```
Comment (AyanSinhaMahapatra) [inline tests/packagedcode/data/build_gradle/groovy/fdroid-version-catalog/gradle/libs.versions.toml:1]: Can you add the link to this file as a comment here?
We need to know the source

Could you also add another real example?

**Review by AyanSinhaMahapatra — CHANGES_REQUESTED** [review]
@Kaushik-Kumar-CEG thanks, see comments for your consideration.

File: src/packagedcode/build_gradle.py (hunk 1 above)
Comment (Kaushik-Kumar-CEG) [reply src/packagedcode/build_gradle.py]: Removed the try/except block, letting exceptions propagate to the upstream catchall

File: src/packagedcode/build_gradle.py (hunk 2 above)
Comment (Kaushik-Kumar-CEG) [reply src/packagedcode/build_gradle.py:358]: added blank lines between logical blocks and ensured proper spacing

File: src/packagedcode/build_gradle.py (hunk 3 above)
Comment (Kaushik-Kumar-CEG) [reply src/packagedcode/build_gradle.py]: changed the logic so unresolved catalog references are kept as is instead of being silently dropped. Only resolved ones get replaced now

File: tests/packagedcode/data/build_gradle/groovy/fdroid-version-catalog/build.gradle-expected.json (hunk 4 above)
Comment (Kaushik-Kumar-CEG) [reply tests/packagedcode/data/build_gradle/groovy/fdroid-version-catalog/build.gradle-expected.json]: added resolved_package for all pinned dependencies. This follows the same pattern used in the npm and nuget handlers.also this adds ~45 lines of PackageData per pinned dep in every expected JSON fixture, which is why the diff looks large (~2800 lines are just regenerated test fixtures)

File: tests/packagedcode/data/build_gradle/groovy/fdroid-version-catalog/gradle/libs.versions.toml (hunk 5 above)
Comment (Kaushik-Kumar-CEG) [reply tests/packagedcode/data/build_gradle/groovy/fdroid-version-catalog/gradle/libs.versions.toml:2]: added source links to both the build.gradle and libs.versions.toml files. also added a second test case from nicbell/material-lists — a real Groovy project that uses version catalogs with the module syntax format. and added support for that TOML format as well

**Comment by Kaushik-Kumar-CEG** [comment]
@AyanSinhaMahapatra 
all review comments have been addressed and force pushed. The large diff is mostly from regenerated test fixtures.adding resolved_package to pinned deps adds ~45 lines per dependency across all existing expected JSON files. actual code changes are around 390 lines
